### PR TITLE
CAN: support more device types in SITL AP_Periph

### PR DESCRIPTION
**Pull request by tridge**
This moves to a multicast UDP method of getting state data from the main simulation ardupilot binary to the simulated peripheral. This means the AP_Periph SITL binary has full state information via the SITL fdm structure and is able to support a much wider range of DroneCAN sensors. In this PR I have enabled:
 - GPS
 - compass
 - airspeed
 - baro
 - rangefinder
 - battery
 - EFI
 - RPM
 - ESC and servo
 - ESC telemetry

This PR also allows for SIM_SPEEDUP other than 1.0 with SITL peripheral, although it will only talk properly to real hw devices if you have SIM_SPEEDUP=1, but for only simulated devices you can do quite high speedups. Time is lock-step synchronised between the peripheral and the ardupilot binary via the timestamp_us field in the multicast UDP state.
@bugobliterator this PR removes the native_millis(), native_micros() etc as I believe these are no longer needed and just complicate the firmware. Please let me know if there is a good reason to keep them. I have tested with bridging to real devices and it works fine as long as you keep SIM_SPEEDUP=1.

Update: now supports ESC and servo peripherals, with feedback coming back by another UDP socket

Update2: can now start a full DroneCAN quadplane with sim_vehicle.py with -f quadplane-can. This includes GPS, airspeed, baro, compass, ESC, servo, ESC telem, rangefinder
Similarly you can choose -f quad-can and get a quadcopter with CAN ESC, CAN ESC telem, compass, baro, GPS, rangefinder

### CODE REVIEW

File: Tools/AP_Periph/AP_Periph.h [hunk 1]
```diff
@@ -371,6 +371,11 @@ class AP_Periph_FW {
         bool locked;
     } uart_monitor;
 #endif
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    SITL::SIM sitl;
+    AP_AHRS ahrs;
```
Comment (peterbarker) [inline Tools/AP_Periph/AP_Periph.h:377]:
```suggestion
#endif
#if AP_AHRS_ENABLED
        AP_AHRS ahrs;
```

... and suitable guards.

File: Tools/AP_Periph/AP_Periph.h [hunk 2]
```diff
@@ -371,6 +371,11 @@ class AP_Periph_FW {
         bool locked;
     } uart_monitor;
 #endif
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
```
Comment (peterbarker) [inline Tools/AP_Periph/AP_Periph.h:375]:
```suggestion
#if AP_SIM_ENABLED
```

File: Tools/AP_Periph/Parameters.cpp [hunk 3]
```diff
@@ -563,6 +563,16 @@ const AP_Param::Info AP_Periph_FW::var_info[] = {
     GOBJECT(rpm_sensor, "RPM", AP_RPM),
 #endif
 
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
```
Comment (peterbarker) [inline Tools/AP_Periph/Parameters.cpp:566]:
```suggestion
#if AP_SIM_ENABLED
```

File: Tools/AP_Periph/Parameters.cpp [hunk 4]
```diff
@@ -563,6 +563,16 @@ const AP_Param::Info AP_Periph_FW::var_info[] = {
...
+    // @Group: SIM_
+    // @Path: ../libraries/SITL/SITL.cpp
+    GOBJECT(sitl, "SIM_", SITL::SIM),
+
+    // @Group: AHRS_
```
Comment (peterbarker) [inline Tools/AP_Periph/Parameters.cpp:571]:
```suggestion
#endif
#if AP_AHRS_ENABLED
    // @Group: AHRS_
```

... similar changes elsewhere.  Will needa line in boards.py enabling it.

File: Tools/ardupilotwaf/boards.py [hunk 5]
```diff
@@ -815,9 +814,17 @@ def configure_env(self, cfg, env):
...
             HAL_PERIPH_ENABLE_GPS = 1,
+            HAL_PERIPH_ENABLE_AIRSPEED = 1,
+            HAL_PERIPH_ENABLE_MAG = 1,
+            HAL_PERIPH_ENABLE_BARO = 1,
+            HAL_PERIPH_ENABLE_RANGEFINDER = 1,
+            HAL_PERIPH_ENABLE_BATTERY = 1,
+            HAL_PERIPH_ENABLE_EFI = 1,
+            HAL_PERIPH_ENABLE_RPM = 1,
+            AP_AIRSPEED_ENABLED = 1,
+            AP_AIRSPEED_AUTOCAL_ENABLE = 0,
```
Comment (peterbarker) [inline Tools/ardupilotwaf/boards.py:827]:
```suggestion
            AP_AIRSPEED_AUTOCAL_ENABLE = 0,
            AP_AHRS_ENABLED = 1,
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: Tools/AP_Periph/AP_Periph.h (hunk 1 above)
Comment (tridge) [reply Tools/AP_Periph/AP_Periph.h:378]: done

File: Tools/AP_Periph/Parameters.cpp (hunk 3 above)
Comment (tridge) [reply Tools/AP_Periph/Parameters.cpp]: done

File: Tools/ardupilotwaf/boards.py (hunk 5 above)
Comment (tridge) [reply Tools/ardupilotwaf/boards.py:827]: done

File: Tools/AP_Periph/AP_Periph.h (hunk 2 above)
Comment (tridge) [reply Tools/AP_Periph/AP_Periph.h]: done

File: Tools/AP_Periph/Parameters.cpp (hunk 4 above)
Comment (tridge) [reply Tools/AP_Periph/Parameters.cpp:572]: done

**Comment by tridge** [comment]
@peterbarker finally got CI passing! The key problem was multicast doesn't work with --force-32bit

File: libraries/SITL/SIM_Gimbal.h [hunk 6]
```diff
@@ -21,7 +21,7 @@
 #include <AP_HAL/AP_HAL_Boards.h>
 
 #ifndef HAL_SIM_GIMBAL_ENABLED
-#define HAL_SIM_GIMBAL_ENABLED (CONFIG_HAL_BOARD == HAL_BOARD_SITL)
+#define HAL_SIM_GIMBAL_ENABLED (CONFIG_HAL_BOARD == HAL_BOARD_SITL) && !defined(HAL_BUILD_AP_PERIPH)
```
Comment (peterbarker) [inline libraries/SITL/SIM_Gimbal.h:24]: This should go into `boards.py` - similarly elsewhere.

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
@peterbarker the flash savings come from the millis/micros plus the inlining of the ESC scaling code. here is elf_diff between the PR for removing native_millis/micros and this PR:
https://uav.tridgell.net/tmp/SITL_Periph_elf_diff/

**Review by peterbarker — APPROVED** [review]
LGeTM

**Pull request merged**